src/dotnet/GeneratorAgent/src/Azure.Tools.GeneratorAgent/GitHubFileService.cs

```diff
@@ -0,0 +1,158 @@
+            ValidationContext validationContext,
+            HttpClient httpClient)
+        {
+            ArgumentNullException.ThrowIfNull(appSettings);
+            ArgumentNullException.ThrowIfNull(logger);
+            ArgumentNullException.ThrowIfNull(validationContext);
+            ArgumentNullException.ThrowIfNull(httpClient);
+
+            AppSettings = appSettings;
+            Logger = logger;
+
+            CommitId = validationContext.ValidatedCommitId;
+            TypespecSpecDir = validationContext.ValidatedTypeSpecDir;
+
+            HttpClient = httpClient;
+            
+            // Log GitHub authentication status
+            bool hasAuthHeader = httpClient.DefaultRequestHeaders.Authorization != null;
+            Logger.LogInformation("GitHub API client initialized. Authentication: {AuthStatus}", 
+                hasAuthHeader ? "Configured" : "Not configured - using rate-limited access");
+        }
+
+        public async Task<Result<Dictionary<string, string>>> GetTypeSpecFilesAsync(CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                Result<Dictionary<string, string>> apiResult = await FetchTypeSpecFilesViaApiAsync(cancellationToken).ConfigureAwait(false);
+                return apiResult;
+            }
+            catch (Exception ex) when (ex is not OperationCanceledException)
+            {
+                Logger.LogCritical(ex, "Unexpected error while fetching TypeSpec files from GitHub repository {Repository}", 
+                    AppSettings.AzureSpecRepository);
+                throw;
+            }
+        }
+
+        private async Task<Result<Dictionary<string, string>>> FetchTypeSpecFilesViaApiAsync(CancellationToken cancellationToken)
+        {
+            try
+            {
+                // GitHub API endpoint for directory contents
+                string apiUrl = $"https://api.github.com/repos/{AppSettings.AzureSpecRepository}/contents/{TypespecSpecDir}?ref={CommitId}";
+
+                HttpResponseMessage response = await HttpClient.GetAsync(apiUrl, cancellationToken).ConfigureAwait(false);
+                
+                if (!response.IsSuccessStatusCode)
+                {
+                    Logger.LogCritical("GitHub API request failed: {StatusCode} {ReasonPhrase}", 
+                        response.StatusCode, response.ReasonPhrase);
+                    throw new HttpRequestException($"GitHub API request failed: {response.StatusCode} {response.ReasonPhrase}");
+                }
+
+                string jsonContent = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
+                GitHubContent[]? contents = JsonSerializer.Deserialize<GitHubContent[]>(jsonContent, JsonOptions);
+
+                if (contents == null)
+                {
+                    throw new InvalidOperationException("Failed to deserialize GitHub API response");
+                }
```

Comment:
nit but it might be nice to make this exception message a little more actionable